core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java

```diff
@@ -1224,10 +1224,18 @@ public Builder builder(RelNode rel, Clause... clauses) {
           needNew = true;
         }
       }
-      if (rel instanceof Aggregate
-          && !dialect.supportsNestedAggregations()
-          && hasNestedAggregations((Aggregate) rel)) {
-        needNew = true;
+
+      if (rel instanceof Aggregate) {
+        Aggregate e = (Aggregate) rel;
+        final boolean nestedAgg = hasNestedAggregations(e);
+        if (withGroupBy() && (!e.getGroupSet().toList().isEmpty()
+            || !nestedAgg || !dialect.supportsNestedAggregations())) {
```

Comment:
Shall we have a method as below:
```
   // Indicate that whether nested aggregation can be applied.
    public boolean canApplyNestedAggregation(Aggregate agg) {
      return agg.getGroupCount() == 0
          && hasNestedAggregations(agg)
          && dialect.supportsNestedAggregations();
    }
```
So the change here will be much simpler that 
```
...
      if (rel instanceof Aggregate
          && this.clauses.contains(Clause.GROUP_BY)
          && !canApplyNestedAggregation((Aggregate) rel)) {
        needNew = true;
      }
...
```

Reply:
The `canApplyNestedAggregation`  method looks like policy but it's actually a confusing mixture of policy & logic. Making it public implies that someone can safely change the policy by overriding the method. At the very least I'd make the method private.

But I think I'd just re-organize that 'if' statement so that it was easier to understand.

Reply:
Does everyone agree that the indentation of lines 1231 and 1232 makes the 'and' / 'or' logic confusion?

Reply:
`if (..` above is confusing.

What is `!e.getGroupSet().toList().isEmpty()`?
Is it equivalent to `isNotGrandTotal(e)`?

Reply:
Well, `!e.getGroupSet().toList().isEmpty()` is certainly equivalent to `!e.getGroupSet().isEmpty()`.

Reply:
-- Update
Sorry, @julianhyde the public there is my typo and my intention was to make the `canApplyNestedAggregation` **private** , I update the method as below
```
    private boolean canApplyNestedAggregation(Aggregate agg) {
      return agg.getGroupCount() == 0
          && hasNestedAggregations(agg)
          && dialect.supportsNestedAggregations();
    }
```
I also think line 1231 and 1232 are very confusing. That's why I propose to use  `canApplyNestedAggregation` instead.
We can apply two `Aggregate` onto a single `SqlSelect` only when it satisfies `canApplyNestedAggregation`, which checks if the outer `agg` is "!isNotGrandTotal"

Reply:
How about this code:
```java
    if (rel instanceof Aggregate) {
        final Aggregate agg = (Aggregate) rel;
        final boolean hasNestedAgg = hasNestedAggregations(agg);
        if (!dialect.supportsNestedAggregations()
            && hasNestedAgg) {
          needNew = true;
        } else {
          if (this.clauses.contains(Clause.GROUP_BY)) {
            // Avoids to lose the distinct attribute of inner agg.
            if (!hasNestedAgg || Aggregate.isNotGrandTotal(agg)) {
              needNew = true;
            }
          }
        }
      }
```

Reply:
> -- Update
> Sorry, @julianhyde the "public" there is a typo and my intention was to make the `canApplyNestedAggregation` **private** , I update the method as below
> 
> ```
>     private boolean canApplyNestedAggregation(Aggregate agg) {
>       return agg.getGroupCount() == 0
>           && hasNestedAggregations(agg)
>           && dialect.supportsNestedAggregations();
>     }
> ```
> 
> I also think line 1231 and 1232 are very confusing. That's why I propose to use `canApplyNestedAggregation` instead.
> We can apply two `Aggregate`s onto a single `SqlSelect` only when it satisfies `canApplyNestedAggregation`, which checks if the outer `agg` is "!isNotGrandTotal"

Hi, like that `!(A and B and C)` does not equal to `!A or !B or !C`, `!canApplyNestedAggregation` does not equal to `(!e.getGroupSet().toList().isEmpty()
            || !nestedAgg || !dialect.supportsNestedAggregations())`, is that right?

Reply:
> How about this code:
> 
> ```java
>     if (rel instanceof Aggregate) {
>         final Aggregate agg = (Aggregate) rel;
>         final boolean hasNestedAgg = hasNestedAggregations(agg);
>         if (!dialect.supportsNestedAggregations()
>             && hasNestedAgg) {
>           needNew = true;
>         } else {
>           if (this.clauses.contains(Clause.GROUP_BY)) {
>             // Avoids to lose the distinct attribute of inner agg.
>             if (!hasNestedAgg || Aggregate.isNotGrandTotal(agg)) {
>               needNew = true;
>             }
>           }
>         }
>       }
> ```
Yeah, it's simpler to my code and it's equivalent to my code.But I think that the logic is still a little bit logic confusing.

Reply:
It doesn't compile (due to invalid type `bool` and reference to `agg` before it is declared), and when I fix the errors, the test fails.

Let's go back to Danny's.